### PR TITLE
Store cost cnstr form

### PR DESCRIPTION
**Pull request by hongkai-dai**
As discussed in https://drakedevelopers.slack.com/archives/C3L92BM2Q/p1492002741266876, the `cost_form` and `cnstr_form` in the constructor of `OptimizationProgram` are unused.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/5870)

### CODE REVIEW

**Comment by hongkai-dai** [comment]
+@soonho-tri  for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-KhzODdUxrVzgL0rsV2Z:bwklt6f)*

**Comment by soonho-tri** [comment]
Sorry for the delays. PTAL.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/linear_program_examples.cc, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiFzPMv0zHgUGH5L-MI:-KiFzPMv0zHgUGH5L-MJ:besytj) ([raw file](https://github.com/robotlocomotion/drake/blob/becdcff2232f3af4397b7b3e87213ab15a12c6ae/drake/solvers/test/linear_program_examples.cc#L133)):*
> ```cpp
> void LinearProgram0::CheckSolution(SolverType solver_type) const {
>   double tol = GetSolverSolutionDefaultCompareTolerance(solver_type);
>   EXPECT_TRUE(CompareMatrices(this->GetSolution(x_), x_expected_, tol,
> ```

FYI, [META], `this->GetSolution(x_)` can be shortened to `GetSolution(x_)`.

---

*[drake/solvers/test/quadratic_program_examples.cc, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiFz4mVYqZaje12AptK:-KiFz4mVYqZaje12AptL:byo9r3n) ([raw file](https://github.com/robotlocomotion/drake/blob/becdcff2232f3af4397b7b3e87213ab15a12c6ae/drake/solvers/test/quadratic_program_examples.cc#L65)):*
> ```cpp
>       x_(),
>       x_expected_(0.25, 0.75) {
>   x_ = this->NewContinuousVariables<2>();
> ```

FYI, [META], all the occurrences of `this->` can be removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-KiFzXE20Z7xJxQoWnig:b-hs4b75)*

**Comment by hongkai-dai** [comment]
No problem. It's not blocking anything. And hope you had fun in HSCC!

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/linear_program_examples.cc, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiFzPMv0zHgUGH5L-MI:-KiHbVyQ-BtlwmANLRA6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/becdcff2232f3af4397b7b3e87213ab15a12c6ae/drake/solvers/test/linear_program_examples.cc#L133)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, [META], `this->GetSolution(x_)` can be shortened to `GetSolution(x_)`.
</blockquote></details>

Done.

---

*[drake/solvers/test/quadratic_program_examples.cc, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiFz4mVYqZaje12AptK:-KiHcBYEM34_Prosoofx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/becdcff2232f3af4397b7b3e87213ab15a12c6ae/drake/solvers/test/quadratic_program_examples.cc#L65)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, [META], all the occurrences of `this->` can be removed.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-KiHcCTAVxf9Cy4sQ051:b-cxny9f)*

**Comment by soonho-tri** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-KiHwzZ4vQOEsSRWlQwK:bnfp4nl)*

**Comment by soonho-tri** [comment]
+@jwnimmer-tri , platform review please. 

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-KiHx1XjL76Yjoknzh2E:b-g8a6a)*

**Comment by jwnimmer-tri** [comment]
+(status: curate commits before merging) 

:lgtm:

---

Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzkZ23kQ2Is_dRgTh:-KiHzkZ23kQ2Is_dRgTi:b-yw8vap) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L32)):*
> ```objc
> void ExpectSolutionCostAccurate(const MathematicalProgram &prog, double tol);
> 
> class OptimizationProgram : public MathematicalProgram {
> ```

FYI I am not a fan of switching from has-a to is-a (this doesn't smell like a subtyping relationship to me), but I guess in a test fixture the rules can bend if you think the brevity is worth it.

---

*[drake/solvers/test/optimization_examples.h, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzTr2PQkR-QuicNI3:-KiHzTr3ni7mNpzrUdtI:b77ed15) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L38)):*
> ```objc
>   OptimizationProgram(CostForm cost_form, ConstraintForm cnstr_form);
> 
>   virtual ~OptimizationProgram() {}
> ```

`override`

---

*[drake/solvers/test/optimization_examples.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzXYmo5q6aXlwveU5:-KiHzXYmo5q6aXlwveU6:b-hyff6d) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L42)):*
> ```objc
>   CostForm cost_form() const {return cost_form_;}
> 
>   ConstraintForm cnstr_form() const {return cnstr_form_;}
> ```

_META_ `cnstr` is an disallowed abbreviation per cppguide

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-KiI-3h889k5s696i36A:bwkpaz3)*

**Comment by hongkai-dai** [comment]
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzkZ23kQ2Is_dRgTh:-KilpoTrMpAkKfn9VG4S:b6bnefr) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L32)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I am not a fan of switching from has-a to is-a (this doesn't smell like a subtyping relationship to me), but I guess in a test fixture the rules can bend if you think the brevity is worth it.
</blockquote></details>

I actually think this is-a mathematical program, with cost, constraints and decision variables. Which part to you think is not a mathematical program?

---

*[drake/solvers/test/optimization_examples.h, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzTr2PQkR-QuicNI3:-KilqNc-ygatbJm5_S3O:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`override`
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzXYmo5q6aXlwveU5:-KilqS-jXhO32450tUcs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_META_ `cnstr` is an disallowed abbreviation per cppguide
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by jwnimmer-tri** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzkZ23kQ2Is_dRgTh:-KilteQzmKpilEu2iAo0:b-uaffb4) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L32)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I actually think this is-a mathematical program, with cost, constraints and decision variables. Which part to you think is not a mathematical program?
</blockquote></details>

Do this callers of this class treat it as a MathematicalProgram?  Do they add new constraints or costs, or call GetSolution?

At first glance, they only call RunProblem, which means OptimizationProgram is a class whose responsibility is to evaluate test case correctness, not a class that models a program.  It _has_ a program under test, it doesn't _offer_ program APIs for others to use.

In other words, as a rule of thumb, for inheritance to be valid, most (or at least _some_) callers must call instances of the derived class via the base class APIs, not only the derived class APIs.

---

*[drake/solvers/test/optimization_examples.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzXYmo5q6aXlwveU5:-KilsFWgk8PIhhKeUP39:b-b3tdyv) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L42)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Ditto for the member field spelling.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzXYmo5q6aXlwveU5:-KiluYn9zo-1VOkOzGGz:b8or3la) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ditto for the member field spelling.
</blockquote></details>

... any many other places in the PR; I didn't tag them all, just grep for `cnstr` everywhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by hongkai-dai** [comment]
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzkZ23kQ2Is_dRgTh:-KilvTPXxl2TnajYKZGQ:b-c8r0kx) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L32)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Do this callers of this class treat it as a MathematicalProgram?  Do they add new constraints or costs, or call GetSolution?

At first glance, they only call RunProblem, which means OptimizationProgram is a class whose responsibility is to evaluate test case correctness, not a class that models a program.  It _has_ a program under test, it doesn't _offer_ program APIs for others to use.

In other words, as a rule of thumb, for inheritance to be valid, most (or at least _some_) callers must call instances of the derived class via the base class APIs, not only the derived class APIs.
</blockquote></details>

All the derived class of `OptimizationProgram` calls `AddCost` `AddConstraint`, and `GetSolution`, such as in https://github.com/RobotLocomotion/drake/pull/5870/files#diff-6fd275dd7296ea4cd4aed7f6cf94865f, is that OK? If this is not a good design I will figure out some other solution.

---

*[drake/solvers/test/optimization_examples.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzXYmo5q6aXlwveU5:-KilzWO_pQJIfxH8-nfy:boygbnj) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

... any many other places in the PR; I didn't tag them all, just grep for `cnstr` everywhere.
</blockquote></details>

Done. Thanks！

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by jwnimmer-tri** [comment]
Reviewed 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/optimization_examples.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzkZ23kQ2Is_dRgTh:-Kim-CVNM91FLIlcbNqU:b-c4p80t) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L32)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

All the derived class of `OptimizationProgram` calls `AddCost` `AddConstraint`, and `GetSolution`, such as in https://github.com/RobotLocomotion/drake/pull/5870/files#diff-6fd275dd7296ea4cd4aed7f6cf94865f, is that OK? If this is not a good design I will figure out some other solution.
</blockquote></details>

... but they could have as well done that on `prog()` like they used to.

For the "As a rule of thumb, for inheritance to be valid, most (or at least some) callers must call instances of the derived class via the base class APIs, not only the derived class APIs" statement, I guess I wasn't counting this-calls.  Inheritance is appropriate when someone upcasts OptimizationProgram to MathematicalProgram, passes around a pointer to MathematicalProgram and then people then do AddCost or something on that.

I think the "some other solution" was what you used to have -- the test fixtures are all working to craft a program that their case class owns, but the test fixture is not a program, it job is to evaluates a solver with respect to a program that it has prepared.  If the "RunProblem" and "CheckSolution" were gone, this would be okay -- there are subclasses that serve to specialize the program with useful sugar.  But since we are adding a fixture wrapper, is no longer is-a _just_ a program, it goes beyond that, thus my intuition for has-a.

In any case, like I said atop, the argument against using inheritance here is that the old way like on master has more boilerplate, so maybe abusing inheritance to reduce boilerplate is worth it for test code.  This the FYI for you to consider, but we we can keep this if you like it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by hongkai-dai** [comment]
Review status: 3 of 7 files reviewed at latest revision, all discussions resolved.

---

*[drake/solvers/test/optimization_examples.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-KiHzkZ23kQ2Is_dRgTh:-KipXwordWEYcHpeBrKt:bop11o5) ([raw file](https://github.com/robotlocomotion/drake/blob/c9ea78de2ebdd93548fa39d4bda1cf55d68989c7/drake/solvers/test/optimization_examples.h#L32)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

... but they could have as well done that on `prog()` like they used to.

For the "As a rule of thumb, for inheritance to be valid, most (or at least some) callers must call instances of the derived class via the base class APIs, not only the derived class APIs" statement, I guess I wasn't counting this-calls.  Inheritance is appropriate when someone upcasts OptimizationProgram to MathematicalProgram, passes around a pointer to MathematicalProgram and then people then do AddCost or something on that.

I think the "some other solution" was what you used to have -- the test fixtures are all working to craft a program that their case class owns, but the test fixture is not a program, it job is to evaluates a solver with respect to a program that it has prepared.  If the "RunProblem" and "CheckSolution" were gone, this would be okay -- there are subclasses that serve to specialize the program with useful sugar.  But since we are adding a fixture wrapper, is no longer is-a _just_ a program, it goes beyond that, thus my intuition for has-a.

In any case, like I said atop, the argument against using inheritance here is that the old way like on master has more boilerplate, so maybe abusing inheritance to reduce boilerplate is worth it for test code.  This the FYI for you to consider, but we we can keep this if you like it.
</blockquote></details>

I see, thanks for the explanation. I reverted back to has-a

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by jwnimmer-tri** [comment]
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870)*

**Comment by hongkai-dai** [comment]
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5870#-:-Kiq5LuY_fxLzdRYWv1U:b-l70wzf)*
